[CORE] Enlarge defaultRecursionLimit by pre-loading the protobuf class

PHILO-HE opened the pull request:
## What changes were proposed in this pull request?

We previously introduced custom protobuf dependency with only defaultRecursionLimit enlarged to avoid the limit is hit for deeply nested plans (one user reported this issue). It is not easy to upgrade this dependency since it requires local build and then publishing the jar to maven repo.

With this change, we can remove the custom protobuf dependency for easily upgrading protobuf when needed.

## How was this patch tested?

Verified by local remote debug.

PHILO-HE commented:
@FelixYBW, @zhztheplayer

zhouyuan commented:
related
https://github.com/apache/incubator-gluten/pull/3814
https://github.com/protocolbuffers/protobuf/issues/13301

FelixYBW commented:
[https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/gluten-iceberg/pom.xml#L84](https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/gluten-iceberg/pom.xml#L84)

set to "provide"  
several other places also need to be updated

FelixYBW commented:
[https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/package/pom.xml#L356](https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/package/pom.xml#L356)  
  
Remove?

PHILO-HE commented:
> https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/gluten-iceberg/pom.xml#L84
> 
> set to "provide" several other places also need to be updated

@FelixYBW, I just removed protobuf from iceberg & hudi module as it is not really used, and changed the scope to provided for hudi module.

PHILO-HE commented:
> https://github.com/apache/incubator-gluten/blob/18f151017f4442b53340802166f7a222b0731723/package/pom.xml#L356
> 
> Remove?

Yes, seems we can remove it. Let's see CI feedback.

FelixYBW commented:
Where is the code to modify protobuf and build customized version? Is it manually upload to mvn repo?

PHILO-HE commented:
> Where is the code to modify protobuf and build customized version? Is it manually upload to mvn repo?

@FelixYBW , the code is maintained in my personal repo: https://github.com/PHILO-HE/protobuf/tree/v3.23.4-0.

Yes, we manually did that.
We firstly obtained a jar by building protobuf from source at local, then go through maven's jar publish process to make it available to download when building gluten.

FelixYBW commented:
> @FelixYBW , the code is maintained in my personal repo: [PHILO-HE/protobuf@`v3.23.4-0`](https://github.com/PHILO-HE/protobuf/tree/v3.23.4-0).
> 
> Yes, we manually did that. We firstly obtained a jar by building protobuf from source at local, then go through maven's jar publish process to make it available to download when building gluten.

Thank you for the PR! manually doing this definitely isn't a right way. Glad we fixed it.

zhztheplayer commented on gluten-core/src/main/scala/org/apache/gluten/GlutenPlugin.scala:
```diff
@@ -267,6 +268,7 @@ private[gluten] class GlutenExecutorPlugin extends ExecutorPlugin {
 
   /** Initialize the executor plugin. */
   override def init(ctx: PluginContext, extraConf: util.Map[String, String]): Unit = {
+    CodedInputStreamClassInitializer
```
Can we give an API for this? E.g.,

```scala
CodedInputStreamClassInitializer.modifyDefaultRecursionLimitUnsafe(100000)
```

PHILO-HE replied:
@zhztheplayer, I feel it may be not necessary to expose the setting API since the value we set is large enough. I prefer hiding the detail in this piece of code. Thanks!

zhztheplayer replied:
Thank you for the explanation. It's OK if we don't pass a number. Though I would suggest having an method for this purpose than relying on class initialization, e.g.,

```
CodedInputStreamClassInitializer.modifyDefaultRecursionLimitUnsafe()
```

Code like `CodedInputStreamClassInitializer` can be less intuitive and perhaps could cause IDEs to complain about the writing or so.

zhztheplayer commented on gluten-core/src/main/scala/org/apache/gluten/initializer/CodedInputStreamClassInitializer.scala:
```diff
@@ -0,0 +1,42 @@
+      val field: Field = clazz.getDeclaredField("defaultRecursionLimit")
+      field.setAccessible(true)
+      // Enlarge defaultRecursionLimit whose original value is 100.
+      field.setInt(null, 100000)
```
nit: We can add a log for such settings.

```
logInfo("... has been modified to ...")
```

PHILO-HE replied:
@zhztheplayer, fixed by https://github.com/apache/incubator-gluten/pull/8996. Thanks!

zhztheplayer reviewed (commented):
Thanks!
